src/main/java/com/mattworzala/canary/server/givemeahome/StructureWriter.java

```diff
@@ -0,0 +1,91 @@
+package com.mattworzala.canary.server.givemeahome;
+
+import com.google.gson.*;
+import com.google.gson.reflect.TypeToken;
+import net.minestom.server.coordinate.Vec;
+import net.minestom.server.instance.block.Block;
+
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class StructureWriter {
+
+    public static void writeStructure(Structure structure, Path filePath) {
+        GsonBuilder gson = new GsonBuilder().excludeFieldsWithoutExposeAnnotation();
+
+        // when registering type adapters for specific generic values
+        // make sure that you register in the order of general -> specific
+
+        Type blockMapType = new TypeToken<Map<Integer, Block>>() {
+        }.getType();
+        gson.registerTypeAdapter(blockMapType, new BlockMapSerializer());
+
+        Type blockDefList = new TypeToken<List<Structure.BlockDef>>() {
+        }.getType();
+        gson.registerTypeAdapter(blockDefList, new BlockDefListSerializer());
+
+        gson.registerTypeAdapter(Vec.class, new VecSerializer());
+
+        var output = gson.create().toJson(structure);
+        File file = filePath.toFile();
+
+        try (FileOutputStream fos = new FileOutputStream(file);
+             BufferedOutputStream bos = new BufferedOutputStream(fos)) {
+            byte[] bytes = output.getBytes();
+            bos.write(bytes);
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        System.out.println("JSON OUTPUT");
+        System.out.println(output);
+    }
+
+    private static class VecSerializer implements JsonSerializer<Vec> {
```

Comment:
Why only define these as serializers? They could be used as deserializers in the reader as well to avoid those static `parseXXX` methods.

Reply:
Gson is not able to deserialize our current format using the higher level api due to the blockmap not being a list of a single type (it is currently a list of string or object). That being said we could probably use the deserializers for the other fields, but I'm not sure what the advantage of calling a specific deserializer is over calling a specific function. If we wanted, we could also change the json format.